plugin/discovery/version_set.go

```diff
@@ -0,0 +1,58 @@
+package discovery
+
+import (
+	version "github.com/hashicorp/go-version"
+)
+
+// A ConstraintsStr is a string containing a possibly-invalid representation
+// of a version constraint provided in configuration. Call Parse on it to
+// obtain a real Constraint object, or discover that it is invalid.
+type ConstraintsStr string
+
+// Parse transforms a ConstraintsStr into a VersionSet if it is
+// syntactically valid. If it isn't then an error is returned instead.
+func (s ConstraintsStr) Parse() (VersionSet, error) {
+	raw, err := version.NewConstraint(string(s))
+	if err != nil {
+		return VersionSet{}, err
+	}
+	return VersionSet{raw}, nil
+}
+
+// MustParse is like Parse but it panics if the constraint string is invalid.
+func (s ConstraintsStr) MustParse() VersionSet {
+	ret, err := s.Parse()
+	if err != nil {
+		panic(err)
+	}
+	return ret
+}
+
+// VersionSet represents a set of versions which any given Version is either
+// a member of or not.
+type VersionSet struct {
+	raw version.Constraints
+}
+
+// Has returns true if the given version is in the receiving set.
+func (s VersionSet) Has(v Version) bool {
+	return s.raw.Check(v.raw)
+}
+
+// Intersection combines the receving set with the given other set to produce a
+// set that is the intersection of both sets, which is to say that it contains
+// only the versions that are members of both sets.
+func (s VersionSet) Intersection(other VersionSet) VersionSet {
+	raw := make(version.Constraints, 0, len(s.raw)+len(other.raw))
+
+	raw = append(raw, s.raw...)
+	raw = append(raw, other.raw...)
```

Comment:
Now that you mention it, I can see why this looks confusing. The type is called "version set", so it's presenting the idea that this is a set of `Version` such that every possible version is either in it or not.

But its *implementation* is the opposite: it's expressed as a list of *constraints*, each of which *removes* versions from the set. So I think `Intersection` is the right noun if you think of this from the perspective of the abstraction (it returns a set of the versions common to both sets) but it feels weird because its implementation is the union of the *constraints*, each of which *removes* versions from the set.

When I was originally putting this together, thinking of it as a positive set of versions rather than a negative set of constraints felt right to me from the perspective of the outward-facing abstraction, since the question you ask is "Is this version allowed" rather than "does this constraint exist"? But your question here makes me wonder it would be best to address this with just a comment clarifying the matter or if it'd be better to just step back from the "set" idea altogether and represent this in a manner more reflective of its implementation.